subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -172,28 +187,46 @@ private String getFilename(Path dir) {
         }
 
         @Override
-        public FileVisitResult visitFile(Path file, @Nullable BasicFileAttributes attrs) {
-            String name = stringInterner.intern(file.getFileName().toString());
-            if (isAllowed(file, name, false, attrs, builder.getRelativePath())) {
-                if (attrs == null) {
-                    throw new FileSnapshottingException(String.format("Cannot read file '%s': not authorized.", file));
-                }
+        public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
+            String name = intern(file.getFileName().toString());
+            if (shouldVisit(file, name, false, attrs, builder.getRelativePath())) {
                 if (attrs.isSymbolicLink()) {
                     // when FileVisitOption.FOLLOW_LINKS, we only get here when link couldn't be followed
                     throw new FileSnapshottingException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", file));
                 }
-                addFileSnapshot(file, name, attrs);
+                builder.visit(snapshotFile(file, name, attrs));
             }
             return FileVisitResult.CONTINUE;
         }
 
+        private FileSystemLocationSnapshot snapshotFile(Path absoluteFilePath, String name, BasicFileAttributes attrs) {
+            String internedAbsoluteFilePath = intern(absoluteFilePath.toString());
+            if (attrs.isRegularFile()) {
+                try {
+                    HashCode hash = hasher.hash(absoluteFilePath.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
+                    FileMetadata metadata = FileMetadata.from(attrs);
+                    return new RegularFileSnapshot(internedAbsoluteFilePath, name, hash, metadata);
+                } catch (UncheckedIOException e) {
+                    LOGGER.info("Could not read file path '{}'.", absoluteFilePath, e);
+                }
+            }
+            return new MissingFileSnapshot(internedAbsoluteFilePath, name);
+        }
+
+        /** unlistable directories (and maybe some locked files) will stop here */
         @Override
         public FileVisitResult visitFileFailed(Path file, IOException exc) {
             // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
             // so we include all the other files apart from the loop.
             // This way, we include each file only once.
-            if (isNotFileSystemLoopException(exc) && isAllowed(file, file.getFileName().toString(), false, null, builder.getRelativePath())) {
-                throw new UncheckedIOException(String.format("Could not read path '%s'.", file), exc);
+            if (isNotFileSystemLoopException(exc)) {
+                String name = intern(file.getFileName().toString());
```

Comment:
IIUC, we haven't interned this before, but we do now. Why?

Why don't we do the interning inside the method? If we need the interned string later, we have `InternableString` for this. Could that be used here?

Reply:
https://github.com/gradle/gradle/pull/9746#discussion_r298256652

@lptr, @wolfs, please decide.

Reply:
Thanks for the context, @paplorinc. I don't feel strongly about whether or not we should intern—I'd leave that up to @wolfs.

Reply:
He already approved, if anything changes, I'll open a new PR

Reply:
I would intern here for consistency. I don't think it will have a big effect, given that the percentage of unavailable files will be very small - if not zero.